### PR TITLE
Notifications counter

### PR DESCRIPTION
**Pull request by sagaidak**

### CODE REVIEW

File: public/env.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// WARN: This is a generated file. Do not modify!
```
Comment (siman) [inline public/env.js]: This file should be in .gitignore

File: src/components/activity/ActivityStream.tsx [hunk 2]
```diff
@@ -82,6 +82,7 @@ export const ViewNotifications = () => {
     if (!myAddress) return;
 
     getNotificationsArray(0).catch(err => new Error(err));
+    clearNotifications(myAddress)
```
Comment (siman) [inline src/components/activity/ActivityStream.tsx]: Here we need to sign a message with accounts private key I guess.

File: src/components/utils/NotifCounter.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+export const NotifCounterProvider = (props: React.PropsWithChildren<{}>) => {
+  const { state: { address: myAddress } } = useMyAccount()
+
+  const [ contextValue, setContextValue ] = useState({ unreadCount: 0 })
+  const [ wsConnected, setWsConnected ] = useState(false)
+  const [ isLoggedIn, setIsLoggedIn ] = useState(false)
+  const [ address, setAddress ] = useState<string | undefined>()
+
+  if (checkIfLoggedIn() && !isLoggedIn) {
```
Comment (siman) [inline src/components/utils/NotifCounter.tsx]: Why do we need this?

File: src/components/utils/NotifCounter.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+export const NotifCounterContext = createContext<NotifCounterContextProps>({ unreadCount: 0 });
+
+export const NotifCounterProvider = (props: React.PropsWithChildren<{}>) => {
+  const { state: { address: myAddress } } = useMyAccount()
+
+  const [ contextValue, setContextValue ] = useState({ unreadCount: 0 })
+  const [ wsConnected, setWsConnected ] = useState(false)
+  const [ isLoggedIn, setIsLoggedIn ] = useState(false)
```
Comment (siman) [inline src/components/utils/NotifCounter.tsx]:
```suggestion
```

File: src/components/utils/NotifCounter.tsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  const [ address, setAddress ] = useState<string | undefined>()
```
Comment (siman) [inline src/components/utils/NotifCounter.tsx]:
```suggestion
```

File: src/components/utils/NotifCounter.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+  useEffect(() => {
+
+    const subscribe = async () => {
+      if (wsConnected || !myAddress) return;
+
+      const ws = new WebSocket(offchainWs)
+
+      ws.onopen = () => {
+        console.log('WS connected (NotificationCounter useEffect)')
```
Comment (siman) [inline src/components/utils/NotifCounter.tsx]:
```suggestion
        console.log('Connected to Notifications Counter Web Socket')

```

File: src/components/utils/NotifCounter.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        ws.send(myAddress?.toString());
+        setWsConnected(true)
+        ws.onmessage = (msg: MessageEvent) => {
+          setContextValue({ unreadCount: msg.data })
+          console.log(msg, 'msg from Effect')
```
Comment (siman) [inline src/components/utils/NotifCounter.tsx]:
```suggestion
          console.log('Received a new value for unread notifications:', unreadCount)
```

File: src/components/utils/NotifCounter.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      if (wsConnected || !myAddress) return;
+
+      const ws = new WebSocket(offchainWs)
+
+      ws.onopen = () => {
+        console.log('WS connected (NotificationCounter useEffect)')
+        ws.send(myAddress?.toString());
+        setWsConnected(true)
+        ws.onmessage = (msg: MessageEvent) => {
+          setContextValue({ unreadCount: msg.data })
```
Comment (siman) [inline src/components/utils/NotifCounter.tsx]:
```suggestion
          const unreadCount = msg.data
          setContextValue({ unreadCount })
```

File: src/components/utils/NotifCounter.tsx [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+          console.log(msg, 'msg from Effect')
+        }
+        ws.onerror = (error) => { console.log('NotificationCounter Websocket Error:', error) }
+      };
+    }
+
+    subscribe()
+  }, [ isLoggedIn, wsConnected ]);
```
Comment (siman) [inline src/components/utils/NotifCounter.tsx]:
```suggestion
  }, [ myAddress?.toString(), wsConnected ]);

```

File: src/layout/SideMenu.tsx [hunk 10]
```diff
@@ -98,6 +107,11 @@ const InnerMenu = () => {
 
   const MenuItems = isLoggedIn ? AuthorizedMenu : DefaultMenu;
 
+  const renderBadge = () => {
+    if (notificationCount <= 0) return null
+    return <Badge count={notificationCount} className="site-badge-count-4" />
```
Comment (siman) [inline src/layout/SideMenu.tsx]:
```suggestion
    return <Badge count={unreadCount} className="site-badge-count-4" />
```

File: src/layout/SideMenu.tsx [hunk 11]
```diff
@@ -98,6 +107,11 @@ const InnerMenu = () => {
 
   const MenuItems = isLoggedIn ? AuthorizedMenu : DefaultMenu;
 
+  const renderBadge = () => {
+    if (notificationCount <= 0) return null
```
Comment (siman) [inline src/layout/SideMenu.tsx]:
```suggestion
    if (!unreadCount || unreadCount <= 0) return null
```

File: src/components/utils/OffchainUtils.ts [hunk 12]
```diff
@@ -34,3 +35,9 @@ export const getNotifications = async (myAddress: string, offset: number, limit:
   const { data } = res;
   return data;
 };
+
+export const clearNotifications = async (myAddress: string): Promise<Activity[]> => {
+  const res = await axios.get(`${offchainUrl}/offchain/clearnotifications/${myAddress}`);
```
Comment (siman) [inline src/components/utils/OffchainUtils.ts]:
```suggestion
  const res = await axios.post(`${offchainUrl}/offchain/notifications/${myAddress}/readAll`);
```

File: src/components/utils/OffchainUtils.ts [hunk 13]
```diff
@@ -6,6 +6,7 @@ const createUrl = (varName: string, defaultUrl: string = 'http://localhost:3001'
 
 export const offchainUrl = createUrl('OFFCHAIN_URL');
 export const ipfsUrl = createUrl('IPFS_URL', getEnv('OFFCHAIN_URL'));
+export const offchainWs = getEnv('OFFCHAIN_WS') as string
```
Comment (siman) [inline src/components/utils/OffchainUtils.ts]: This `getEnv()` should return a string already, why do we need a cast to `as string`?
```suggestion
export const offchainWs = getEnv('OFFCHAIN_WS') as string
```

File: src/components/utils/OffchainUtils.ts [hunk 14]
```diff
@@ -34,3 +35,9 @@ export const getNotifications = async (myAddress: string, offset: number, limit:
   const { data } = res;
   return data;
 };
+
+export const clearNotifications = async (myAddress: string): Promise<Activity[]> => {
+  const res = await axios.get(`${offchainUrl}/offchain/clearnotifications/${myAddress}`);
+  const { data } = res;
```
Comment (siman) [inline src/components/utils/OffchainUtils.ts]: Seems like this function doesn't need to return any result.

File: src/layout/SideMenu.tsx [hunk 15]
```diff
@@ -23,12 +24,20 @@ interface MenuItem {
 const InnerMenu = () => {
   const { toggle, state: { collapsed, trigerFollowed } } = useSidebarCollapsed();
   const { state: { address: myAddress } } = useMyAccount();
+  const { unreadCount } = useNotifCounter()
   const isLoggedIn = checkIfLoggedIn();
   const [ followedBlogsData, setFollowedBlogsData ] = useState([] as BlogData[]);
   const [ loaded, setLoaded ] = useState(false);
+  const [ notificationCount, setNotificationCount ] = useState(unreadCount)
```
Comment (siman) [inline src/layout/SideMenu.tsx]:
```suggestion
```

File: src/layout/SideMenu.tsx [hunk 16]
```diff
@@ -23,12 +24,20 @@ interface MenuItem {
...
   const router = useRouter();
   const { pathname } = router;
 
+  useEffect(() => {
```
Comment (siman) [inline src/layout/SideMenu.tsx]: Delete this `useEffect`. It's not needed here at all.

File: src/components/utils/NotifCounter.tsx (hunk 3 above)
Comment (sagaidak) [reply src/components/utils/NotifCounter.tsx]: We should reconnect WS if account was changed.

File: src/components/utils/NotifCounter.tsx (hunk 4 above)
Comment (sagaidak) [reply src/components/utils/NotifCounter.tsx]: Same here. We should try to connect to websocket only after user login. Maybe there is more elegant solution?

File: src/components/utils/NotifCounter.tsx (hunk 5 above)
Comment (sagaidak) [reply src/components/utils/NotifCounter.tsx]: Here we should trigger websocket connection after user has changed his account

File: src/components/utils/OffchainUtils.ts (hunk 14 above)
Comment (sagaidak) [reply src/components/utils/OffchainUtils.ts]: But if there is some error?

File: src/components/utils/OffchainUtils.ts [hunk 17]
```diff
@@ -34,3 +35,15 @@ export const getNotifications = async (myAddress: string, offset: number, limit:
...
+
+export const clearNotifications = async (myAddress: string): Promise<void> => {
+  try {
+    const res = await axios.post(`${offchainUrl}/offchain/notifications/${myAddress}/readAll`);
+
+    if (res.status !== 200) {
+      console.warn('Failed to mark all notifications as read for account:', myAddress, 'res.status:', res.status)
+    }
+  } catch (err) {
+    console.log('Failed to mark all notifications as read for account:', err)
```
Comment (siman) [inline src/components/utils/OffchainUtils.ts]:
```suggestion
    console.log('Failed to mark all notifications as read for account: ${myAddress}', err)
```

**Pull request merged**